### PR TITLE
test(api): refactor e2e tests

### PR DESCRIPTION
**Pull request by hardcoretime**
## Description

Add stage 1: apply virtualization resources
Update ginkgo version: 2.20.0
## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: tests/e2e/testdata/virtualization-resources/vm/base/cfg/cloudinit.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  - name: cloud
+    # passwd: cloud
+    passwd: $6$rounds=4096$vln/.aPHBOI7BMYR$bBMkqQvuGs5Gyd/1H5DP4m9HjQSy.kgrxpaGEHwkX7KEFV8BS.HZWPitAtZ2Vd8ZqIZRqmlykRCagTgPejt1i.
+    shell: /bin/bash
+    sudo: ALL=(ALL) NOPASSWD:ALL
+    chpasswd: { expire: False }
+    lock_passwd: false
+    ssh_authorized_keys:
+      # testcases
+      - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQC9bSJ6RpHDB0jc8oSHk4kgEaoNzv3vNqMhNosLpXYVE6R7CHO3OjOK3eXulVcgiMCRH6EGcDOkMXGM2lvSr5t5HfnqZ1bGxGRCUtMeYs8zDSCMMOuyb57LNjwNAD7cjNhTGvGqFSXqoH5P6lHGIT0M82lU8jo0BIurcnjMgiFU8w== testcases
```
Comment (fl64) [inline tests/e2e/testdata/virtualization-resources/vm/base/cfg/cloudinit.yaml:12]: JFY this key must be generated before testing can begin

**Comment by fl64** [comment]
Please remove all commented lines

**Review by fl64 — DISMISSED** [review]

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/kubectl/kubectl.go [hunk 2]
```diff
@@ -228,7 +244,19 @@ func (k KubectlCMD) WaitResource(resource Resource, name string, opts WaitOption
...
-		timeout = opts.Timeout
+		timeout = opts.Timeout * time.Second
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	return k.ExecContext(ctx, cmd)
+}
+
+func (k KubectlCMD) WaitResources(resource Resource, names []string, opts WaitOptions) *executor.CMDResult {
+	namesList := strings.Join(names, " ")
```
[the remark this answers is not in the document]
Comment (yaroslavborbat) [reply tests/e2e/kubectl/kubectl.go:255]: namesList := strings.Join(names, " ")

File: tests/e2e/config/config.go [hunk 3]
```diff
@@ -127,3 +208,92 @@ func (c *Config) setEnvs() error {
...
+		return err
+	}
+
+	filePath := "/tmp/mc.yaml"
+	writeErr := os.WriteFile(filePath, updatedModuleConfig, 0644)
+	if writeErr != nil {
+		return writeErr
+	}
+
+	res := kubectl.Apply(filePath, kc.ApplyOptions{})
```
Comment (yaroslavborbat) [inline tests/e2e/config/config.go:278]: I don't want this

File: tests/e2e/git/git.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
```
Comment (yaroslavborbat) [inline tests/e2e/git/git.go:18]: Why do you need this?

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/default_config.yaml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-namespace: "test-d8-virtualization"
+namespace: "testcases"
```
Comment (yaroslavborbat) [inline tests/e2e/default_config.yaml:1]: rename field namespace

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/config/config.go [hunk 6]
```diff
@@ -127,3 +137,28 @@ func (c *Config) setEnvs() error {
...
+		log.Fatal(unmarshalErr)
+	}
+
+	kustomizeFile.Namespace = namespace
+	updatedKustomizeFile, marshalErr := yamlv3.Marshal(&kustomizeFile)
+	if marshalErr != nil {
+		log.Fatal(updatedKustomizeFile)
+	}
+
+	writeErr := os.WriteFile(filePath, updatedKustomizeFile, 0644)
```
[the remark this answers is not in the document]
Comment (hardcoretime) [reply tests/e2e/config/config.go:248]: all files will be copied to /tmp before changing. files in repository will not be changed.

File: tests/e2e/testdata/virtualization-resources/vm/overlays/custom-ip/vmip.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+apiVersion: virtualization.deckhouse.io/v1alpha2
+kind: VirtualMachineIPAddress
+metadata:
+  name: vm
+spec:
+  type: Static
+  staticIP: 10.255.255.254
```
[the remark this answers is not in the document]
Comment (hardcoretime) [reply tests/e2e/testdata/virtualization-resources/vm/overlays/custom-ip/vmip.yaml:7]: it is required for kustomization. this IP will be replaced with IP address from module config subnet.

File: tests/e2e/config/config.go [hunk 8]
```diff
@@ -43,13 +67,58 @@ func GetConfig() (*Config, error) {
...
-	Ipam             IpamConf         `yaml:"ipam"`
-	HelperImages     HelperImages     `yaml:"helperImages"`
+	Namespace               string           `yaml:"namespace"`
+	ClusterTransport        ClusterTransport `yaml:"clusterTransport"`
+	Disks                   DisksConf        `yaml:"disks"`
+	VM                      VmConf           `yaml:"vm"`
+	Ipam                    IpamConf         `yaml:"ipam"`
+	HelperImages            HelperImages     `yaml:"helperImages"`
+	VirtualizationResources string           `yaml:"virtualizationResources"`
+	CustomSubnet            string           `yaml:"customSubnet"`
```
[the remark this answers is not in the document]
Comment (hardcoretime) [reply tests/e2e/config/config.go:121]: removed.

File: tests/e2e/config/config.go [hunk 9]
```diff
@@ -127,3 +208,92 @@ func (c *Config) setEnvs() error {
...
+	return slices.Contains(mc.Spec.Settings.VirtualMachineCIDRs, subnet)
+}
+
+func SetupCustomSubnet(mc *ModuleConfig, subnet string) error {
+	subnetConfigured := CheckCustomSubnet(mc, subnet)
+	if subnetConfigured {
+		return nil
+	}
+
+	mc.Spec.Settings.VirtualMachineCIDRs = append(mc.Spec.Settings.VirtualMachineCIDRs, conf.CustomSubnet)
```
[the remark this answers is not in the document]
Comment (hardcoretime) [reply tests/e2e/config/config.go:266]: this operation removed.

File: tests/e2e/testdata/virtualization-resources/cvi/cvi-from-vi-alpine-http.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+kind: ClusterVirtualImage
+metadata:
+  name: cvi-from-vi-alpine-http
+spec:
+  dataSource:
+    type: ObjectRef
+    objectRef:
+      kind: VirtualImage
+      name: vi-alpine-http
+      namespace: testcases
```
[the remark this answers is not in the document]
Comment (hardcoretime) [reply tests/e2e/testdata/virtualization-resources/cvi/cvi-from-vi-alpine-http.yaml:12]: this required for kustomization.

File: tests/e2e/util_test.go [hunk 11]
```diff
@@ -202,3 +210,111 @@ func ChmodFile(pathFile string, permission os.FileMode) {
...
+			}
+
+			if res.WasSuccess() && res.StdOut() == "" {
+				return intToIP(ip).String(), nil
+			}
+
+		}
+	}
+	return "", fmt.Errorf("error: cannot find unassigned IP address")
+}
```
Comment (yaroslavborbat) [inline tests/e2e/util_test.go:284]:
```suggestion
func FindUnassignedIP(subnets []string) (string, error) {
	var metas []metav1.ObjectMeta
	res := kubectl.List(kc.ResourceVMIPLease, kc.GetOptions{Output: "jsonpath='{.items[*].metadata}'"})
	if !res.WasSuccess() {
		return "", fmt.Errorf("failed to get vmipl: %s", res.StdErr())
	}
	if err := yamlv3.Unmarshal(res.StdOutBytes(), &metas); err != nil {
		return "", fmt.Errorf("failed to unmarshal vmipls metadata: %w", err)
	}
	reservedIPs := make(map[string]struct{}, len(metas))
	for _, m := range metas {
		ip := strings.ReplaceAll(strings.TrimPrefix(m.Name, "ip-"), "-", ".")
		reservedIPs[ip] = struct{}{}
	}
	for _, rawSubnet := range subnets {
		ip, subnet, err := net.ParseCIDR(rawSubnet)
		if err != nil {
			return "", fmt.Errorf("failed to parse subnet %s: %w", rawSubnet, err)
		}
		start := ipToInt(ip.To4())
		mask := net.IP(subnet.Mask).To4()
		broadcast := start | ^(ipToInt(mask))
		for uip := broadcast - 1; uip > start+1; uip-- {
			ip := intToIP(uip).String()
			if _, found := reservedIPs[ip]; found {
				continue
			}
			return ip, nil
		}
	}
	return "", fmt.Errorf("error: cannot find unassigned IP address")
}
```

I I didn't test suggestion func

File: tests/e2e/util_test.go [hunk 12]
```diff
@@ -202,3 +210,111 @@ func ChmodFile(pathFile string, permission os.FileMode) {
...
+
+func intToIP(ipInt uint32) (result net.IP) {
+	result = net.IPv4(
+		byte(ipInt>>24),
+		byte(ipInt>>16),
+		byte(ipInt>>8),
+		byte(ipInt),
+	)
+	return
+}
```
Comment (yaroslavborbat) [inline tests/e2e/util_test.go:258]: Are you sure about these functions? I don't fully understand how they work, leave comments

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: tests/e2e/util_test.go (hunk 11 above)
Comment (yaroslavborbat) [reply tests/e2e/util_test.go:284]: although your way may be better in some cases

**Review by yaroslavborbat — DISMISSED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**